### PR TITLE
Adds Razer Huntsman V2 Tenkeyless Support

### LINKED ISSUE #1681: Support for Razer Huntsman V2 Tenkeyless
_Supporting the Huntsman V2 (full-size) https://github.com/openrazer/openrazer/pull/1674 first would probably be best before adding this TKL version._ − **It's done**.

Will be fixed by #1754

### Device Info

* Name: Razer Huntsman V2 Tenkeyless
* Product Number: RZ03-03940600-R3F1
* Model Number: RZ03-0394

### Device Information
```
Bus 001 Device 004: ID 1532:026b Razer USA, Ltd Razer Huntsman V2 Tenkeyless
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0
  bDeviceSubClass         0
  bDeviceProtocol         0
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x026b
  bcdDevice            1.01
  iManufacturer           1 Razer
  iProduct                2 Razer Huntsman V2 Tenkeyless
  iSerial                 0
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          0
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     177
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      79
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      0
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      22
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval               1
```

```
lrwxrwxrwx 1 root root  10 24 nov.  14:16 usb-Razer_Razer_Huntsman_V2_Tenkeyless-event-if01 -> ../event17
lrwxrwxrwx 1 root root  10 24 nov.  14:16 usb-Razer_Razer_Huntsman_V2_Tenkeyless-event-if04 -> ../event19
lrwxrwxrwx 1 root root   9 24 nov.  14:16 usb-Razer_Razer_Huntsman_V2_Tenkeyless-event-kbd -> ../event6
lrwxrwxrwx 1 root root   9 24 nov.  14:16 usb-Razer_Razer_Huntsman_V2_Tenkeyless-if01-event-kbd -> ../event7
lrwxrwxrwx 1 root root  10 24 nov.  14:16 usb-Razer_Razer_Huntsman_V2_Tenkeyless-if02-event-mouse -> ../event18
lrwxrwxrwx 1 root root   9 24 nov.  14:16 usb-Razer_Razer_Huntsman_V2_Tenkeyless-if02-mouse -> ../mouse3
```

### Packet Captures
…

### Synapse Screenshots
tkl01
tkl02

### LINKED ISSUE #1750: Razer Huntsman v2 Tenkeyless - Not recognized
Model Name: Razer Huntsman v2 Tenkelyless (TKL) (Tournamnet Edition)
Model Number: RZ03-03941600-R3W1

### PR DESCRIPTION
**Pull request by saibotk**
Currently based on #1674 so do not merge until this is in master.

Fixes #1681
Fixes #1750

## TODO:
- [x] Check media / special controls
- [x] Fix keyboard layout mappings in custom color view
- [x] Better device image

### CODE REVIEW

**Comment by ThomChroma** [comment]
>     * [ ]  Better device image

From the [support page](https://mysupport.razer.com/app/answers/detail/a_id/5638):
https://dl.razerzone.com/src/5638/5638-1-en-v1.png

**Comment by z3ntu** [comment]
FYI #1674 has been merged

**Comment by ciwin** [comment]
Hi, any chance to proceed with this PR? I have the Razer Huntsman V2 Tenkeyless and would love to use it with Linux. Until now it is not recognized (1532:026B). If I can help to speed things up, let me know! Thanks!

**Comment by timkister** [comment]
Hey, I just bought the Keyboard and I‘m using Arch Linux. If I can help to speed up things let me know! Thanks!

**Comment by gageracer** [comment]
How can we fix the build issue? Is there anything I can help with?

**Comment by j-zehnder** [comment]
Possibly related to #1810?

**Comment by saibotk** [comment]
Hey all,

sorry had some pretty busy weeks and no chance to make any progress.

I might get to this tomorrow morning and try to get this up to date.

There is not much help needed i think. The only thing would be to test it once its up to date.

And im still unsure about my initial remarks of the wrong layout in the key customization screen. I dont really know what i need to touch or change to fix this.

**Comment by gageracer** [comment]
> Hey all,
> 
> sorry had some pretty busy weeks and no chance to make any progress.
> 
> I might get to this tomorrow morning and try to get this up to date.
> 
> There is not much help needed i think. The only thing would be to test it once its up to date.
> 
> And im still unsure about my initial remarks of the wrong layout in the key customization screen. I dont really know what i need to touch or change to fix this.

Would love to test it

**Comment by saibotk** [comment]
This now is in line with the changes that ware made to the huntsman v2 addition.
It can now be tested :)

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 1]
```diff
@@ -475,6 +475,27 @@ class RazerBlackWidowXChroma(_RippleKeyboard):
...
+    """
+    Class for the Razer Huntsman V2 Tenkeyless
+    """
+    EVENT_FILE_REGEX = re.compile(r'.*Razer_Huntsman_V2_Tenkeyless(-if01)?-event-kbd')
+
+    USB_VID = 0x1532
+    USB_PID = 0x026B
+    HAS_MATRIX = True
+    DEDICATED_MACRO_KEYS = False
+    MATRIX_DIMS = [6, 22]
```
Comment (j-zehnder) [inline daemon/openrazer_daemon/hardware/keyboards.py]: I am not sure on the MATRIX_DIMS here, the other tenkeyless keyboard is set to 6x18, not 6x22

Comment (saibotk) [reply daemon/openrazer_daemon/hardware/keyboards.py]: Good point i missed that when adjusting for the tkl

**Comment by timkister** [comment]
Build works and so far no major problems :) Can I help with the keyboard Layout fixing?

**Comment by jugglerchris** [comment]
Hi,
I've just got a Huntsman V2 TKL, and it works pretty well with the driver from this branch.  Things I noticed that didn't seem quite right:
* In Polychromatic, the "Reactive" mode doesn't work (the button doesn't do anything)
* Macro recording doesn't work, but I'm not sure if it's expected to?  When I press Fn-M the F9/M key flashes red a few times.  The help text in Polychromatic says it should work with the M-keys, but there aren't any M-keys - but the instruction leaflet doesn't mention such a restriction.  (This is my first gaming keyboard, so I'm not sure what to expect!)

Thanks for your work!

**Comment by z3ntu** [comment]
@saibotk Could you please check for the Reactive effect? If it doesn't work - and you can't get it to work - then please remove the functionality for this device.

@jugglerchris Macro support in OpenRazer is quite limited and we're also looking to remove it in the future because of this. You can use https://github.com/sezanzeb/input-remapper for macros and remapping and these things.

**Comment by Zwygi49** [comment]
I installed today based on the suggestion from @z3ntu the support from @saibotk repository on Debian Bullseye and it seems to work.. after a few hours I had no troubles, brightness and colors settings work, all effects I have tested so fat seems to work fine (didn't test Reactive), as I'm not a gamer, I didn't test game [effects
Huntsman_V2_Tenkeyless_Polichromatic
Will come back as soon as I have something to add or mention. 
Thanks for your support !

**Comment by saibotk** [comment]
Sorry for the delay, had been pretty busy with university. But i will check the effect later and get this thing ready to go @z3ntu 

Also thank you @Zwygi49 for testing :)

**Comment by saibotk** [comment]
This is now ready to merge/review

Its up to date with the latest changes and features. 
Also i fixed the reactive effect 🙌🤩

**Comment by saibotk** [comment]
The only thing I'm unsure about, is whether this could also be reduced to a Matrix of (6, 17) since in polychromatic I can create an effect and the right most column does not affect any buttons at all.

**Comment by 0pLuS0** [comment]
HI, @saibotk 

I just cloned from your repo today;

https://github.com/saibotk/openrazer.git

I'm running Slackware 15, kernel 5.15.36, openrazer 3.4.0 kernel, polychromatic 0.7.3.

My Huntsman V2 TKL shows up unrecognised. I'm not sure the status on the repo, if it should work, but I thought I should post this in case it should, and this helps, and of course, I'd love to get my keyboard working.

THANKS

openrazer

**Comment by saibotk** [comment]
Are you using the correct branch? Since for me it does work and the keyboard shows up every time.

**Comment by blackduckx** [comment]
Hello @saibotk, I just wanted to let you know that your implementation support just looks to work like a charm :)

Test env: Debian Testing

Tested with polychromatic (After applying the 250Hz fix https://github.com/polychromatic/polychromatic/pull/391),
all effects + Game Mode did just work fine.

Still can't / do not know how to check the polling Frequency ;)

Regarding your remark on the 6, 17 matrix, I also noticed the last column did no have any effect in Polychromatics

For those who want to test on Debian:

```
git clone git@github.com:saibotk/openrazer.git
cd openrazer
git checkout feature_huntsman_v2_tkl
./scripts/build_debs.sh
cd dist
sudo dpkg -i openrazer-daemon_3.4.0-0_all.deb openrazer-driver-dkms_3.4.0-0_all.deb python3-openrazer_3.4.0-0_all.deb openrazer-meta_3.4.0-0_all.deb

# And REBOOT 
```

**Comment by blackduckx** [comment]
@0pLuS0 

Run this before installing (And do not forget to reboot)
```
git checkout feature_huntsman_v2_tkl
```

**Comment by 0pLuS0** [comment]
Ok, I left that out, so now I did;

git clone https://github.com/saibotk/openrazer.git
cd openrazer
git checkout feature_huntsman_v2_tkl

And then I compiled it from a build script as;(This is only the build cmds from the script)

make python_library_install DESTDIR=$PKG
make daemon_install DESTDIR=$PKG
make udev_install DESTDIR=$PKG

Then I rebooted, and in Slackware I don't use any login UI, I just log in from the console. When I tried to start typing my user/pass, the keyboard now won't type. Hmm

When I looked in the log, there was no entry for the keyboard, only my RazerBasiliskV2.

2022-08-14 12:34:49 | razer                          | INFO     | Initialising Daemon (v3.4.0). Pid: 1402
2022-08-14 12:34:49 | razer.screensaver              | INFO     | Initialising DBus Screensaver Monitor
2022-08-14 12:34:49 | razer                          | INFO     | Found device.0: 0003:1532:0085.0001
2022-08-14 12:34:49 | razer.device0                  | INFO     | Initialising device.0 RazerBasiliskV2
2022-08-14 12:34:49 | razer                          | INFO     | Serving DBus

I also have this in /etc/rc.d/rc.modules.local; (for the keyboard module to load at boot up)

/sbin/modprobe razerkbd

**Comment by 0pLuS0** [comment]
I actually made a mistake and used saibotk's repo, but I only used it for the dameon, and forgot to use it for the kernel as well.

Now the keyboard shows up in Polychromatic.

But there's an error with it, it's still not working properly.

HuntsmanV2Error

This is what the details shows.
```
Traceback (most recent call last):
  File ’/usr/lib64/python3.9/site-packages/polychromatic/middleman.py’, line 165, in get_device
    device = module.get_device(uid)
  File ’/usr/lib64/python3.9/site-packages/polychromatic/backends/openrazer.py’, line 607, in get_device
    ’id’: ids[rate],
KeyError: 250
```
Slackware 15.0 uses python3-3.9.12 and I'm running Kernel 5.19.2, which I noticed has HID_RAZER [=m], I compiled as a module, not sure if this module helps any, just thought I'd mention it.

The log shows;
```
2022-08-18 15:07:57 | razer                          | INFO     | Initialising Daemon (v3.4.0). Pid: 1422
2022-08-18 15:07:57 | razer.screensaver              | INFO     | Initialising DBus Screensaver Monitor
2022-08-18 15:07:57 | razer                          | INFO     | Found device.0: 0003:1532:0085.0001
2022-08-18 15:07:57 | razer.device0                  | INFO     | Initialising device.0 RazerBasiliskV2
2022-08-18 15:07:57 | razer                          | INFO     | Found device.1: 0003:1532:026B.0006
2022-08-18 15:07:57 | razer.device1                  | INFO     | Initialising device.1 RazerHuntsmanV2Tenkeyless
2022-08-18 15:07:57 | razer.device1                  | INFO     | Putting device into driver mode. Daemon will handle special functionality
2022-08-18 15:07:57 | razer                          | INFO     | Serving DBus
```

**Comment by saibotk** [comment]
Polychromatic cannot show this device yet since it does not support all settings, you will need to use polychromatic from the backend rework branch.

**Comment by blackduckx** [comment]
@0pLuS0 Use this patch on the current polychromatic version : https://github.com/polychromatic/polychromatic/pull/391

**Comment by lah7** [comment]
Or, simply add the missing `250` to the version you have installed (v0.7.3), see here: https://github.com/polychromatic/polychromatic/issues/430#issuecomment-1209927560

This will be fixed in the next release.

**Comment by 0pLuS0** [comment]
YEAH, SUCCESS! 👍 

polychromatic

THANKS everyone!

**Comment by 0pLuS0** [comment]
I'm Dual Boot Win11/Slack, the settings don't stick on Dual Boot.

In the screen shot above, you'll notice I'm using Static, but if I boot into Windows, then later boot into Slack, I have to open up Poylchromatic, then click as an example 'None' then 'Static' and I'm back to my settings working, otherwise the colors are rotating around. 

As for everything else working, the keyboard seems fine, for now it just seems like the Effects don't stick, if I boot into Windows, then Slack.

If I simply reboot Slackware, everything works ok, so the only problem is between booting from Windows into Slack.

hmm ;/

**Comment by lah7** [comment]
@0pLuS0 Seems expected. Razer Synapse on Windows controls everything in "software", so when you reboot/power cycle after Windows, the keyboard is still in that "software" effect. 

**Solution A:** Fully uninstall Razer Syanpse and set the hardware effect via OpenRazer.

**Solution B:** Check `~/.config/openrazer/razer.conf` and set `restore_persistence = True` which will re-apply the chosen hardware effect after logging in on Linux (but not during boot)

**Comment by 0pLuS0** [comment]
I do have it on; ```restore_persistence = True``` this isn't working when I log in. I had this same issue with a RazerBlackWidowV3 with openrazer 3.3.0 and when I enabled this, it worked, but it's not working with this Huntsman.

Also I've noticed, when I reboot from Win11 back to Slack, I'll see in the console saying it failed to load the module razerkbd, but If I only reboot Slack, and boot back into Slack, I don't see this. hmm odd

I have in /etc/rc.d/rc.modules.local;
```/sbin/modprobe razerkbd```

Should I have the module entry in /etc/rc.d, is it needed?

**Comment by 0pLuS0** [comment]
I ran into another issue, when I started up my computer for the day, I booted first, straight into Slackware. When I logged in and was on my desktop, I opened polychromatic and clicked on None then Static, this time it didn't work and the colors kept changing on the keyboard. So I started to click on the other effects, breath, spectrum, then back to none, then static. Even after clicking on all these effects, the colors still kept changing.

After clicking on the effects several times, it finally stopped and stayed on Static, which is what I use, on the default razer green at 35% brightness.

**Comment by lwndow** [comment]
Wanted to add my support for this working on Nobara/Fedora 36 when using @saibotk branch and manually adding the 250hz polling to Poly. Excellent work! :+1:

**Comment by z3ntu** [comment]
@saibotk Please make sure you have the "Allow edits by maintainers" checkbox ticked on this PR, otherwise I can't merge it.

(note to myself, pushed rebase as 8b7ff56)

**Comment by saibotk** [comment]
@z3ntu It was already on ;)

**Comment by z3ntu** [comment]
Are you sure? I'm getting an error trying to push to that branch (which works for all other PRs)
```
ERROR: Permission to saibotk/openrazer.git denied to z3ntu.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Otherwise please just force push the commit 8b7ff56 to this branch yourself :shrug:

**Comment by saibotk** [comment]
@z3ntu weird, i triple checked, it is set to allow.

Anyways, i squashed the commits and force pushed

**Comment by rbelando** [comment]
When this PR is planned to be merge?

**Comment by z3ntu** [comment]
> @z3ntu weird, i triple checked, it is set to allow.

Manualy applied as aaf3ca4a then. Thanks for the patch!

**Pull request closed without merging**